src/protocol/attribution/apply_attribution_window.rs

```diff
@@ -0,0 +1,351 @@
+                value,
+            )
+        })
+        .collect::<Vec<_>>())
+}
+
+/// Computes time deltas from each trigger event to its nearest matching source event.
+///
+/// # Errors
+/// Fails if the multiplication fails.
+async fn prefix_sum_time_deltas<F, C, T>(
+    ctx: &C,
+    input: &[MCApplyAttributionWindowInputRow<F, T>],
+) -> Result<Vec<T>, Error>
+where
+    F: Field,
+    C: Context<F, Share = T>,
+    T: Arithmetic<F>,
+{
+    let num_rows = input.len();
+
+    // Pre-compute `is_trigger_bit * helper_bit'.
+    // Later in the prefix-sum loop, this vector is updated in each iteration to help
+    // accumulate timedeltas and determine when to stop accumulating.
+    let stop_bit_context = ctx
+        .narrow(&Step::IsTriggerBitTimesHelperBit)
+        .set_total_records(num_rows - 1);
+    let mut stop_bits = Some(T::ZERO)
+        .into_iter()
+        .chain(
+            try_join_all(input.iter().skip(1).enumerate().map(|(i, x)| {
+                let c = stop_bit_context.clone();
+                let record_id = RecordId::from(i);
+                async move {
+                    c.multiply(record_id, &x.is_trigger_report, &x.helper_bit)
+                        .await
+                }
+            }))
+            .await?,
+        )
+        .collect::<Vec<_>>();
+
+    // First, create a vector of timedeltas. This vector contains non-zero values only for
+    // rows with `stop_bit` = 1, meaning that the row is a trigger event, and has the same
+    // match key as the event one above.
+    let t_delta_context = ctx
+        .narrow(&Step::InitializeTimeDelta)
+        .set_total_records(num_rows - 1);
+    let mut t_delta = Some(T::ZERO)
+        .into_iter()
+        .chain(
+            try_join_all(
+                zip(input.iter(), input.iter().skip(1))
+                    .zip(stop_bits.iter().skip(1))
+                    .enumerate()
+                    .map(|(i, ((prev, curr), b))| {
+                        let c = t_delta_context.clone();
+                        let record_id = RecordId::from(i);
+                        let delta = curr.timestamp.clone() - &prev.timestamp;
+                        async move { c.multiply(record_id, &delta, b).await }
+                    }),
+            )
+            .await?,
+        )
+        .collect::<Vec<_>>();
+
+    // Do the prefix-sum accumulation of timedeltas. This is similar to
+    // `do_the_binary_tree_thing()`, but it sums up values in top-down order.
+    for (depth, step_size) in std::iter::successors(Some(1_usize), |prev| prev.checked_mul(2))
+        .take_while(|&v| v < num_rows)
+        .enumerate()
+    {
+        let record_count = num_rows - step_size;
+
+        let depth_i_ctx = ctx
+            .narrow(&InteractionPatternStep::from(depth))
+            .set_total_records(record_count);
+        let mt_ctx = depth_i_ctx.narrow(&Step::PreviousStopBitTimesCurrent);
+
+        let mut futures = Vec::with_capacity(record_count);
```

Comment:
I don't think you need a vector here, `try_join_all` takes an iterator, so you can feed it one created from `0..record_count` range